### ISSUE TITLE
[no-await-in-loop] ignores the await that may not be executed

### ISSUE THREAD

**Issue by regseb** [eslint/eslint#13304]
**What rule do you want to change?**
**[no-await-in-loop](https://eslint.org/docs/rules/no-await-in-loop)**. I suggest not to produce warnings when the `await` may not be executed. Because:

- it's in the *consequent* or *alternate* of a `if`;
- it's in the *consequent* or *alternate* of a conditional (ternary) operator;
- it's to the *right* of a logical expression;
- there is a `continue` (not in sub-loop) before the `await`;
- there is a `break` (not in sub-loop or `switch`) in the loop;
- there is a `return` in the loop.

**Does this change cause the rule to produce more or fewer warnings?**
Fewer.

**How will the change be implemented? (New option, new default behavior, etc.)?**
The rule takes a new option `"ignore"` with the value :

- `"none"` (default): reports all the `await`s in a loop (same behavior as now);
- `"maybe"`: indicates only the `await`s that are always called.

**Please provide some example code that this change will affect:**

👍 Examples of correct code for this rule with the `{ "ignore": "maybe" }` options:

```js
// Comments are CSS selector like.
// LoopStatement = ForStatement[test|update|body]
//                 ForOfStateMent[body]|ForInStatement[body]|
//                 WhileStatement[test|body]|DoWhileStatement[test|body]
(async () => {
    // LoopStatement
    //   IfStatement[consequent|alernate]
    //     AwaitExpression
    for (const i of [1, 2, 3]) {
        if (1 === i) {
            await Promise.resolve(i);
        }
    }

    // LoopStatement
    //   ConditionalExpression[consequent|alternate]
    //     AwaitExpression
    for (const i of [1, 2, 3]) {
        1 === i ? await Promise.resolve(i)
                : 0;
    }

    // LoopStatement
    //   LogicalExpression[right]
    //     AwaitExpression
    for (const i of [1, 2, 3]) {
        if (1 === i || 2 === await Promise.resolve(i)) {
            // Do something.
        }
    }

    // LoopStatement
    //   *:has(ContinueStatement /* not in sub-loop */)
    //   ~ AwaitExpression
    for (const i of [1, 2, 3]) {
        if (1 === i) {
            continue;
        }
        await Promise.resolve(i);
    }

    // LoopStatement:has(AwaitExpression):has(BreakStatement /* not in sub-loop or switch */)
    for (const i of [1, 2, 3]) {
        await Promise.resolve(i);
        break;
    }

    // ForOfStatement:has(AwaitExpression):has(ReturnStatement)
    for (const i of [1, 2, 3]) {
        await Promise.resolve(i);
        return;
    }
})();
```

👎 Examples of incorrect code for this rule with the `{ "ignore": "maybe" }` options:

```js
(async () => {
    for (const i of [1, 2, 3]) {
        await Promise.resolve(i);
    }

    for (const i of [1, 2, 3]) {
        if (2 === await Promise.resolve(i)) {
            // Do something.
        }
    }

    for (const i of [1, 2, 3]) {
        for (const j of [4, 5, 6]) {
            continue;
        }
        await Promise.resolve(i);
    }

    for (const i of [1, 2, 3]) {
        await Promise.resolve(i);
        if (1 === i) {
            continue;
        }
    }

    for (const i of [1, 2, 3]) {
        await Promise.resolve(i);
        for (const j of [4, 5, 6]) {
            break;
        }
    }
})();
```

**What does the rule currently do for this code?**
The _correct code_ produces warnings: _Unexpected \`await\` inside a loop._

**What will the rule do after it's changed?**
With:

- `2`: the _correct code_ produces warnings;
- `[2, { "ignore": "none" }]`: the _correct code_ produces warnings;
- `[2, { "ignore": "maybe" }]`: the _correct code_ doesn't produce warning;

**Are you willing to submit a pull request to implement this change?**
No.

**Comment by kaicataldo** [comment]
I think this option goes against the intention of the rule, because the loop iteration will still be blocked by the awaited async function. If all the values aren't asynchronous, it seems like it would be better to do something like the following (contrived example):

```js
let counter = 0;

function fnGenerator() {
  return Math.random() < 0.5 
    ? counter++
    : new Promise((resolve, reject) => {
      setTimeout(() => {
        resolve(counter++);
      }, Math.random() * 100)
    });
}

(async function() {
  const values = new Array(10).fill(fnGenerator);
  const resolvedValues = await Promise.all(values.map(v => Promise.resolve(v())));
  console.log(resolvedValues);
})();
```

To me, this seems like a case where it would be better to either disable the rule for the loop or disable the rule entirely.

**Comment by regseb** [comment]
If the promises are heavy and not all of them are necessary: it can be counterproductive to execute all promises.

In these two examples, the first does not execute unnecessary requests; the second sends all requests.

```javascript
(async () => {
    async function findFirst() {
        const langs = ["en", "fr", "es", "ja", "de", "ru", "it", "zh", "pt", "pl"];
        for (const lang of langs) {
            const response = await fetch(`https://${lang}.wikipedia.org`);
            if (response.ok) {
                return lang;
            }
        }
        return null;
    }
})();
```

```javascript
(async () => {
    async function findFirst() {
        const fetches = ["en", "fr", "es", "ja", "de", "ru", "it", "zh", "pt", "pl"]
                .map((lang) => fetch(`https://${lang}.wikipedia.org`));
        for await (const response of fetches) {
            if (response.ok) {
                return lang;
            }
        }
        return null;
    }
})();
```

It's a shame to have to manually manage each loop by disabling this rule when it could be automated.

**Comment by kaicataldo** [comment]
Thanks for the extra example - that helped clarify the ask for me a bit. Let's see what the rest of the team thinks. Please note that we'll need a community member to implement this if it's accepted (you can see our rule change policy [here](https://eslint.org/blog/2020/05/changes-to-rules-policies)).
